File: omok/netspec.py
"""Backend-independent description of the policy/value network.

The canonical parameter names below are exactly the PyTorch ``state_dict``
keys; the MLX backend translates to and from them.  That makes a checkpoint
portable: train on a CUDA box, resume on a Mac, export from either.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any

import numpy as np

from .encode import NUM_PLANES


@dataclass(frozen=True)
class NetSpec:
    board_size: int = 15
    in_planes: int = NUM_PLANES
    channels: int = 96
    blocks: int = 6
    policy_channels: int = 4
    value_channels: int = 2
    value_hidden: int = 128

    @property
    def action_size(self) -> int:
        return self.board_size * self.board_size

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "NetSpec":
        fields = {f for f in cls.__dataclass_fields__}
        return cls(**{k: v for k, v in data.items() if k in fields})

    @classmethod
    def from_config(cls, cfg) -> "NetSpec":
        return cls(
            board_size=cfg.game.board_size,
            in_planes=NUM_PLANES,
            channels=cfg.net.channels,
            blocks=cfg.net.blocks,
            policy_channels=cfg.net.policy_channels,
            value_channels=cfg.net.value_channels,
            value_hidden=cfg.net.value_hidden,
        )

    def parameter_count(self) -> int:
        return sum(int(np.prod(shape)) for _, shape in parameter_shapes(self))


def _bn_shapes(name: str, channels: int) -> list[tuple[str, tuple[int, ...]]]:
    return [
        (f"{name}.weight", (channels,)),
        (f"{name}.bias", (channels,)),
        (f"{name}.running_mean", (channels,)),
        (f"{name}.running_var", (channels,)),
    ]


def parameter_shapes(spec: NetSpec) -> list[tuple[str, tuple[int, ...]]]:
    c = spec.channels
    n = spec.board_size
    shapes: list[tuple[str, tuple[int, ...]]] = []
    shapes.append(("stem.conv.weight", (c, spec.in_planes, 3, 3)))
    shapes += _bn_shapes("stem.bn", c)
    for i in range(spec.blocks):
        shapes.append((f"blocks.{i}.conv1.weight", (c, c, 3, 3)))
        shapes += _bn_shapes(f"blocks.{i}.bn1", c)
        shapes.append((f"blocks.{i}.conv2.weight", (c, c, 3, 3)))
        shapes += _bn_shapes(f"blocks.{i}.bn2", c)
    pc, vc = spec.policy_channels, spec.value_channels
    shapes.append(("policy.conv.weight", (pc, c, 1, 1)))
    shapes += _bn_shapes("policy.bn", pc)
    shapes.append(("policy.fc.weight", (spec.action_size, pc * n * n)))
    shapes.append(("policy.fc.bias", (spec.action_size,)))
    shapes.append(("value.conv.weight", (vc, c, 1, 1)))
    shapes += _bn_shapes("value.bn", vc)
    shapes.append(("value.fc1.weight", (spec.value_hidden, vc * n * n)))
    shapes.append(("value.fc1.bias", (spec.value_hidden,)))
    shapes.append(("value.fc2.weight", (1, spec.value_hidden)))
    shapes.append(("value.fc2.bias", (1,)))
    return shapes


def init_weights(spec: NetSpec, seed: int = 0) -> dict[str, np.ndarray]:
    """Canonical initial weights, identical on every backend for a given seed."""
    rng = np.random.default_rng(seed)
    weights: dict[str, np.ndarray] = {}
    for name, shape in parameter_shapes(spec):
        module = name.rsplit(".", 2)[-2]
        leaf = name.rsplit(".", 1)[-1]
        if leaf == "running_mean":
            weights[name] = np.zeros(shape, dtype=np.float32)
        elif leaf == "running_var":
            weights[name] = np.ones(shape, dtype=np.float32)
        elif module.startswith("bn"):
            fill = np.ones if leaf == "weight" else np.zeros
            weights[name] = fill(shape, dtype=np.float32)
        elif module.startswith("conv"):
            fan_in = int(np.prod(shape[1:]))
            weights[name] = rng.normal(0.0, np.sqrt(2.0 / fan_in), size=shape).astype(np.float32)
        elif leaf == "weight":  # fully connected
            bound = 1.0 / np.sqrt(shape[1])
            weights[name] = rng.uniform(-bound, bound, size=shape).astype(np.float32)
        else:  # fully connected bias
            weights[name] = np.zeros(shape, dtype=np.float32)
    return weights


def check_weights(spec: NetSpec, weights: dict[str, np.ndarray]) -> None:
    expected = dict(parameter_shapes(spec))
    missing = sorted(set(expected) - set(weights))
    extra = sorted(set(weights) - set(expected))
    if missing or extra:
        raise ValueError(f"weight mismatch: missing={missing[:4]} extra={extra[:4]}")
    for name, shape in expected.items():
        if tuple(weights[name].shape) != tuple(shape):
            raise ValueError(f"{name}: expected {shape}, got {tuple(weights[name].shape)}")

File: rust/crates/omok-core/src/netspec.rs
//! Backend-independent description of the policy/value network.
//!
//! The parameter names are the PyTorch `state_dict` keys used by the trainer,
//! so a checkpoint written on any of its backends loads here unchanged.
//! Mirrors `omok/netspec.py`.

use crate::encode::NUM_PLANES;
use crate::npz::Weights;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct NetSpec {
    pub board_size: usize,
    pub in_planes: usize,
    pub channels: usize,
    pub blocks: usize,
    pub policy_channels: usize,
    pub value_channels: usize,
    pub value_hidden: usize,
}

impl Default for NetSpec {
    fn default() -> Self {
        NetSpec {
            board_size: 15,
            in_planes: NUM_PLANES,
            channels: 96,
            blocks: 6,
            policy_channels: 4,
            value_channels: 2,
            value_hidden: 128,
        }
    }
}

impl NetSpec {
    pub fn action_size(&self) -> usize {
        self.board_size * self.board_size
    }

    /// Read a spec out of a checkpoint's sidecar JSON (`{"spec": {...}}`) or
    /// out of a bare spec object; missing fields keep their defaults.
    pub fn from_json(value: &serde_json::Value) -> Self {
        let spec = value.get("spec").unwrap_or(value);
        let mut out = NetSpec::default();
        let mut read = |key: &str, into: &mut usize| {
            if let Some(v) = spec.get(key).and_then(|v| v.as_u64()) {
                *into = v as usize;
            }
        };
        read("board_size", &mut out.board_size);
        read("in_planes", &mut out.in_planes);
        read("channels", &mut out.channels);
        read("blocks", &mut out.blocks);
        read("policy_channels", &mut out.policy_channels);
        read("value_channels", &mut out.value_channels);
        read("value_hidden", &mut out.value_hidden);
        out
    }

    /// Recover the spec from the weight shapes alone, so a `.npz` with no
    /// sidecar JSON still loads.
    pub fn from_weights(weights: &Weights) -> Result<Self, String> {
        let stem = weights
            .get("stem.conv.weight")
            .ok_or("weights have no stem.conv.weight; is this an Omok checkpoint?")?;
        if stem.shape.len() != 4 {
            return Err(format!("stem.conv.weight has shape {:?}, expected 4 axes", stem.shape));
        }
        let channels = stem.shape[0];
        let in_planes = stem.shape[1];
        let blocks = (0..).take_while(|i| weights.contains_key(&format!("blocks.{i}.conv1.weight"))).count();

        let policy_conv = weights.get("policy.conv.weight").ok_or("weights have no policy head")?;
        let value_conv = weights.get("value.conv.weight").ok_or("weights have no value head")?;
        let policy_channels = policy_conv.shape[0];
        let value_channels = value_conv.shape[0];

        let policy_fc = weights.get("policy.fc.weight").ok_or("weights have no policy.fc.weight")?;
        let action_size = policy_fc.shape[0];
        let board_size = (action_size as f64).sqrt().round() as usize;
        if board_size * board_size != action_size {
            return Err(format!("policy head has {action_size} outputs, which is not a square board"));
        }
        let value_hidden = weights
            .get("value.fc1.weight")
            .ok_or("weights have no value.fc1.weight")?
            .shape[0];

        Ok(NetSpec {
            board_size,
            in_planes,
            channels,
            blocks,
            policy_channels,
            value_channels,
            value_hidden,
        })
    }

    pub fn parameter_count(&self, weights: &Weights) -> usize {
        let _ = self;
        weights.values().map(|t| t.len()).sum()
    }

    pub fn describe(&self) -> String {
        format!(
            "{}x{} board, {} blocks x {} channels",
            self.board_size, self.board_size, self.blocks, self.channels
        )
    }
}
